src/interpreter/stepper.rs

```diff
@@ -0,0 +1,214 @@
+            assert!(self.fncx.frame().constants.is_empty());
+            ConstantExtractor {
+                span: stmt.span,
+                mir: &self.mir,
+                gecx: self.fncx.gecx,
+                frame: self.fncx.stack.last_mut().expect("stack empty"),
+            }.visit_statement(block, stmt);
+            if self.fncx.frame().constants.is_empty() {
+                self.process = Self::statement;
+                return Ok(Event::Assignment);
+            } else {
+                self.process = Self::constant;
+                return Ok(Event::Constant);
+            }
+        }
+
+        let terminator = basic_block.terminator();
+        assert!(self.fncx.frame().constants.is_empty());
+        ConstantExtractor {
+            span: terminator.span,
+            mir: &self.mir,
+            gecx: self.fncx.gecx,
+            frame: self.fncx.stack.last_mut().expect("stack empty"),
+        }.visit_terminator(block, terminator);
+        if self.fncx.frame().constants.is_empty() {
+            self.process = Self::terminator;
+            Ok(Event::Terminator)
+        } else {
+            self.process = Self::constant;
+            Ok(Event::Constant)
+        }
+    }
+
+    /// returns the statement that will be processed next
+    pub fn stmt(&self) -> &mir::Statement {
+        &self.fncx.basic_block().statements[self.fncx.frame().stmt]
+    }
+
+    /// returns the terminator of the current block
+    pub fn term(&self) -> &mir::Terminator {
+        self.fncx.basic_block().terminator()
+    }
+
+    pub fn block(&self) -> mir::BasicBlock {
+        self.fncx.frame().next_block
+    }
+}
+
+struct ConstantExtractor<'a, 'b: 'mir, 'mir: 'a, 'tcx: 'b> {
+    span: Span,
+    mir: &'a CachedMir<'mir, 'tcx>,
+    frame: &'a mut Frame<'mir, 'tcx>,
+    gecx: &'a mut GlobalEvalContext<'b, 'tcx>,
+}
+
+impl<'a, 'b, 'mir, 'tcx> ConstantExtractor<'a, 'b, 'mir, 'tcx> {
+    fn static_item(&mut self, def_id: DefId, substs: &'tcx subst::Substs<'tcx>, span: Span) {
+        let cid = ConstantId::Static {
+            def_id: def_id,
+            substs: substs,
+        };
+        if self.gecx.statics.contains_key(&cid) {
+            return;
+        }
+        let mir = self.gecx.load_mir(def_id);
+        let ptr = self.gecx.alloc_ret_ptr(mir.return_ty, substs).expect("there's no such thing as an unreachable static");
+        self.gecx.statics.insert(cid.clone(), ptr);
+        self.frame.constants.push((cid, span, ptr, mir));
+    }
+}
+
+impl<'a, 'b, 'mir, 'tcx> Visitor<'tcx> for ConstantExtractor<'a, 'b, 'mir, 'tcx> {
+    fn visit_constant(&mut self, constant: &mir::Constant<'tcx>) {
+        self.super_constant(constant);
+        match constant.literal {
+            // already computed by rustc
+            mir::Literal::Value { .. } => {}
+            mir::Literal::Item { def_id, substs } => {
+                let item_ty = self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs);
```

Comment:
Was that from `gecx.monomorphize(constant.ty)`? Cause if it was just the result of `lookup_item_type`, that's not enough.
Generally speaking, `constant.ty == self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs)` but the latter is more effort.